hazelcast/src/main/java/com/hazelcast/sql/impl/worker/QueryFragmentExecutable.java

```diff
@@ -0,0 +1,218 @@
+    ) {
+        this.stateCallback = stateCallback;
+        this.arguments = arguments;
+        this.exec = exec;
+        this.inboxes = inboxes;
+        this.outboxes = outboxes;
+        this.fragmentPool = fragmentPool;
+    }
+
+    public Collection<Integer> getInboxEdgeIds() {
+        return inboxes.keySet();
+    }
+
+    public Collection<Integer> getOutboxEdgeIds() {
+        return outboxes.keySet();
+    }
+
+    /**
+     * Add operation to be processed.
+     */
+    public void addOperation(QueryAbstractExchangeOperation operation) {
+        operationCount.incrementAndGet();
+        operations.addLast(operation);
+    }
+
+    public void run() {
+        try {
+            if (completed) {
+                return;
+            }
+
+            // Setup the executor if needed.
+            setupExecutor();
+
+            // Feed all batches to relevant inboxes first. Set the upper boundary on the number of batches to avoid
+            // starvation when batches arrive quicker than we are able to process them.
+            int maxOperationCount = operationCount.get();
+            int processedBatchCount = 0;
+
+            QueryOperation operation;
+
+            while ((operation = operations.pollFirst()) != null) {
+                if (operation instanceof QueryBatchExchangeOperation) {
+                    QueryBatchExchangeOperation operation0 = (QueryBatchExchangeOperation) operation;
+
+                    InboundHandler inbox = inboxes.get(operation0.getEdgeId());
+                    assert inbox != null;
+
+                    InboundBatch batch = new InboundBatch(
+                        operation0.getBatch(),
+                        operation0.isLast(),
+                        operation0.getCallerId()
+                    );
+
+                    inbox.onBatch(batch, operation0.getRemainingMemory());
+                } else {
+                    assert operation instanceof QueryFlowControlExchangeOperation;
+
+                    QueryFlowControlExchangeOperation operation0 = (QueryFlowControlExchangeOperation) operation;
+
+                    Map<UUID, OutboundHandler> edgeOutboxes = outboxes.get(operation0.getEdgeId());
+                    assert edgeOutboxes != null;
+
+                    OutboundHandler outbox = edgeOutboxes.get(operation.getCallerId());
+                    assert outbox != null;
+
+                    outbox.onFlowControl(operation0.getRemainingMemory());
+                }
+
+                if (++processedBatchCount >= maxOperationCount) {
+                    break;
+                }
+            }
+
+            operationCount.addAndGet(-1 * processedBatchCount);
+
+            IterationResult res = exec.advance();
+
+            // Send flow control messages if needed.
+            if (res != IterationResult.FETCHED_DONE) {
+                for (InboundHandler inbox : inboxes.values()) {
+                    inbox.sendFlowControl();
```

Comment:
Do we have to send flow control if `res` is `WAIT`?

Reply:
Yes. E.g. if we are fetching a stream of rows from the remote operator, then we will return `WAIT` for all batches except for the last one. Nevertheless, flow control should work during this process as usual.